program/src/matching.rs

```diff
@@ -1044,16 +1167,37 @@ impl<'a> Book<'a> {
         // generate new order id
         let order_id = market.gen_order_id(Side::Bid, price);
 
-        // if post only and price >= best_ask, return
         // Iterate through book and match against this new bid
+        //
+        // Any changes to matching asks are collected in ask_changes
+        // and then applied after this loop.
         let mut rem_quantity = quantity; // base lots (aka contracts)
-        while rem_quantity > 0 {
-            let best_ask_h = match self.get_best_ask_handle() {
-                None => break,
-                Some(h) => h,
-            };
+        let mut ask_changes: Vec<(NodeHandle, i64)> = vec![];
+        let mut ask_deletes: Vec<i128> = vec![];
+        let mut number_of_dropped_expired_orders = 0;
+        for (best_ask_h, best_ask) in self.asks.iter_all_including_invalid() {
+            if rem_quantity <= 0 {
+                break;
+            }
+
+            if !best_ask.is_valid(now_ts) {
+                // Remove the order from the book unless we've done that enough
+                if number_of_dropped_expired_orders < DROP_EXPIRED_ORDER_LIMIT {
+                    number_of_dropped_expired_orders += 1;
+                    let event = OutEvent::new(
```

Comment:
question: should expired orders end up on queue? Update: I see that booted out orders are also on queue, so lgtm

Reply:
still I wonder how does one figure out that is an expired event? should we create a new type? ExpiredEvent?

Reply:
hmm OutEvent is also ok I guess